### ISSUE TITLE
Initialize slow

### ISSUE THREAD

**Issue by MrKumaPants** [kogosoftwarellc/open-api#372]
I'm loading a very large schema. It has about 500 paths, and about 1200 external references. I'm calling `initialize` like so:

```
openapi.initialize({
	apiDoc: this.apiDoc,
    app,
    enableObjectCoercion: true,
    externalSchemas: this.externalSchemas,
    paths: this.paths,
    promiseMode: true,
    validateApiDoc: true
});
```

`apiDoc` and `externalSchemas` are loaded from yaml files and sent to `initialize` as objects. One microservice uses about 34 paths. However it takes a consistent 26 seconds for initialize to finish.

Is there any way to speed this up on my end, in the library, etc?

**Comment by jsdevel** [comment]
dang!  500 paths is a lot!

are the external references downloaded when your server starts up?  if so, i'd bet that's where most of your slowness is happening.

**Comment by MrKumaPants** [comment]
Yeah, it is very large. We have microservices for different parts of the schema.

I do a `yaml.safeLoad(file)` for them all before sending them to `initialize.externalSchemas`. Reading the yaml files into JS objects only takes a few seconds.

**Comment by MrKumaPants** [comment]
Edit: Ran a traffic monitor on my server. Looks like despite me supplying externalSchemas it's still downloading them from the web. I wonder if it's taking the full ref into account and not stripping the ref path?

**Comment by jsdevel** [comment]
well, with 500 paths, i can only imagine how many operations you have.  if you were to `require` all those files without calling `.initialize`, how long does it take?

**Comment by MrKumaPants** [comment]
This is all done loading yaml files and programmatically generating the path operations. Loading and generating all of that only takes 4 seconds or so.

**Comment by jsdevel** [comment]
right, but during `.initialize`, it's going to call `require` for each of the paths right?

what's the output of this?

```
find ./paths-dir -type f | wc -l
```

**Comment by jsdevel** [comment]
depending on how many require calls the framework has to make, it may make sense to use `args.operations`.

**Comment by MrKumaPants** [comment]
I'm sending the paths to initialize as:

```
[
    {
        path,
        module: { get: Function() }
    }
]
```

**Comment by jsdevel** [comment]
oh, well that changes everything lol

using `console`, can you log before and after times before and after the initialize call?

**Comment by MrKumaPants** [comment]
Using this:

```
        console.log(new Date().toTimeString());
        openapi.initialize({
            apiDoc: this.apiDoc,
            app,
            enableObjectCoercion: true,
            externalSchemas: this.externalSchemas,
            paths: this.paths,
            promiseMode: true,
            validateApiDoc: true
        });
        console.log(new Date().toTimeString());
```
Gives this:
```
18:57:13 GMT-0600 (Central Standard Time)
...
18:57:39 GMT-0600 (Central Standard Time)
```

**Comment by jsdevel** [comment]
wow.  well, that's definitely slow.  it would be nice if we had a way to optionally turn on profiling logs

**Comment by jsdevel** [comment]
i can offer theories, but it would be amazing if you could profile this locally: https://nodejs.org/en/docs/guides/simple-profiling/

i assume it's as slow or slower on your local machine?

**Comment by MrKumaPants** [comment]
```
   ticks parent  name   2573    9.4%  v8::internal::JsonStringifier::SerializeString(v8::internal::Handle<v8::internal::String>)   2573  100.0%    v8::internal::Builtin_JsonStringify(int, v8::internal::Object**, v8::internal::Isolate*)   2561   99.5%      LazyCompile: *stringify /opt/exten/ms-volume/node_modules/@exten/mw-lib/node_modules/fast-json-stable-stringify/index.js:19:32   2554   99.7%        LazyCompile: *stringify ./node_modules/fast-json-stable-stringify/index.js:19:32   2509   98.2%          LazyCompile: *stringify ./node_modules/fast-json-stable-stringify/index.js:19:32   2496   99.5%            LazyCompile: *stringify ./node_modules/fast-json-stable-stringify/index.js:19:32
```

**Comment by jsdevel** [comment]
https://github.com/kogosoftwarellc/open-api/blob/master/packages/openapi-framework/src/util.ts#L150

**Comment by jsdevel** [comment]
That copy function is actually called a lot: https://github.com/kogosoftwarellc/open-api/blob/master/packages/openapi-framework/index.ts

The point was to guarantee a deep come in case the rest of the framework alerted the original doc in some way.

**Comment by jsdevel** [comment]
@MrKumaPants is this a deal breaker for you, or just an annoyance?

**Comment by MrKumaPants** [comment]
I hacked around it by traversing the schema/extensions and load only the ones i need. However, the constant encode and decode to JSON could be fixed and most of the slowdown is in ajv.

**Comment by jsdevel** [comment]
> ... the constant encode and decode to JSON could be fixed and most of the slowdown is in ajv.

ok.  thanks for doing some recon.  if you can think of a way to avoid the encode/decode to create a fresh clone i'm all ears.  perhaps we can use immutable.js?

**Comment by jsau-** [comment]
Sorry to revive an old issue thread - I'll probably be opening a dedicated one for this once I've dug enough to come up with some concrete suggestions on how to improve, but I thought it'd be worth leaving here in case anyone sees in the meantime.

I've found that a _lot_ of the startup time I see in large specs is specifically coupling high numbers of paths, with high numbers of components, _and_ using response validation middleware.

I wrote a [small sample project](https://github.com/jsau-/express-openapi-memory-usage-response-validation) to profile memory utilization off the back of some issues we've been having with our larger services, and specifically combining all three of those things makes things _massively_ slower to startup, and eats a huge chunk of memory ([some sample findings in the README see heap usage grow by 5-10x depending on number of paths/components](https://github.com/jsau-/express-openapi-memory-usage-response-validation#sample-findings).)

So TL/DR if you're seeing slow startup times for very large specs and currently use response validation middleware, probably worth considering disabling response validation middleware entirely in prod using `'x-express-openapi-disable-response-validation-middleware': true`.

**Comment by ralphpig** [comment]
I've looked into this issue a few times over the last couple years (40+ second startup time) and I think I've finally got it figured out.

For each route handler / operation, an `OpenAPIRequestValidator` is created. Each `OpenAPIRequestValidator` instance creates it's own `Ajv` instance, and calls `ajv.addSchema()` for **all** schemas in the API doc. This also includes `customFormats`, `customKeywords`, `componentSchemas`, and `externalSchemas`. This is the cause of slow `initialize` because it's re-initializing all schemas for every single route / operation.

So the solution is to use a singleton Ajv instance for **all** `OpenAPIRequestValidator`, and only init the global schemas one time. This is also recommended by [Ajv](https://ajv.js.org/guide/managing-schemas.html#compiling-during-initialization), as it can cache compiled schemas across all routes

I've copy/pasted the provided `OpenAPIRequestValidator` and changed it to init a singleton Ajv instance, and only init the global args once (`componentSchemas`, `customFormats`, etc). In my case, this reduce the `initialize()` time from **40s** to **3s**.  I'm also seeing a near 70% drop in idle memory usage (~700MB to ~200MB), which I can attribute to no longer creating hundreds of Ajv instances, each storing hundreds of the same schemas.

The same improvement can be done for `OpenAPIResponseValidator` too.

---

@jsdevel

I'm more than happy to submit an PR for this, but would like your input on design / approach. I'm not sure of the best way to support a singleton Ajv across multiple `OpenAPIRequestValidator` while ensuring backwards compat and fitting your design style.

Some thoughts:

- Maybe the best approach be to include a `OpenAPIRequestValidatorSingleton` that implements the same functionality as `OpenAPIRequestValidator`, however uses an Ajv singleton. Then consumers can specify that `openapi-framework` use this class if they need to. And these same for the response validator.

- Another easy approach would be to add a flag to `OpenAPIRequestValidator` like `ajvUseSingleton: boolean`, that's default `false`. Then `OpenAPIRequestValidator` can manage it's own singleton, and this should be backwards compat. However I can imagine it's weird for a library to create it's own singleton that shared across multiple instances. This might be okay for now though consumer of `OpenAPIRequestValidator` already has no option to control the Ajv instance behind the scenes.

- Another good approach, but is likely breaking and not worth doing, would be that `OpenAPIRequestValidator` is provided with an initialized `Ajv` with all global schemas, and `OpenAPIRequestValidator` is only tasked with adding per-route validations. I'd imagine `openapi-framework` would have to be responsible for initializing Ajv, which I'm not sure you'd want that. This also might break direct consumers of `OpenAPIRequestValidator`
